### PR TITLE
[SPARK-33045][SQL][FOLLOWUP] Support built-in function like_any and fix StackOverflowError issue.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Spark already support `LIKE ANY` syntax, but it will throw `StackOverflowError` if there are many elements(more than 14378 elements). We should implement built-in function for LIKE ANY to fix this issue.

Why the stack overflow can happen in the current approach ?
The current approach uses reduceLeft to connect each `Like(e, p)`, this will lead the the call depth of the thread is too large, causing `StackOverflowError` problems.

Why the fix in this PR can avoid the error?
This PR support built-in function for `LIKE ANY` and avoid this issue.

### Why are the changes needed?
1.Fix the `StackOverflowError` issue.
2.Support built-in function `like_any`.

### Does this PR introduce _any_ user-facing change?
'No'.

### How was this patch tested?
Jenkins test.

### CODE REVIEW

**Comment by beliefer** [comment]
cc @wangyum @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 1]
```diff
@@ -255,11 +265,67 @@ abstract class LikeAllBase extends UnaryExpression with ImplicitCastInputTypes w
...
+
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val eval = child.genCode(ctx)
+    val patternClass = classOf[Pattern].getName
+    val javaDataType = CodeGenerator.javaType(child.dataType)
+    val pattern = ctx.freshName("pattern")
+    val valueArg = ctx.freshName("valueArg")
+    val patternCache = ctx.addReferenceObj("patternCache", cache.asJava)
+
+    val checkMatchCode = if (isNotSpecified) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:296]: this can be put in the base trait as well.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 2]
```diff
@@ -1404,11 +1404,24 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
       case SqlBaseParser.LIKE =>
         Option(ctx.quantifier).map(_.getType) match {
           case Some(SqlBaseParser.ANY) | Some(SqlBaseParser.SOME) =>
-            getLikeQuantifierExprs(ctx.expression).reduceLeft(Or)
+            validate(!ctx.expression.isEmpty, "Expected something between '(' and ')'.", ctx)
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1407]: It seems we do not need empty check in Line#1387 now: https://github.com/apache/spark/pull/30465/files#diff-26fe6b511f4b8aedf5bdffae0d7c03aba69463c09fa4f881d77520125796e0faR1387

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 3]
```diff
@@ -1404,11 +1404,24 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
...
+            val expressions = expressionList(ctx.expression)
+            if (expressions.size > SQLConf.get.optimizerMultiLikeConversionThreshold &&
+              expressions.forall(_.foldable) && expressions.forall(_.dataType == StringType)) {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1410]: indentation?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -216,16 +216,16 @@ object SQLConf {
...
 
-  val OPTIMIZER_LIKE_ALL_CONVERSION_THRESHOLD =
-    buildConf("spark.sql.optimizer.likeAllConversionThreshold")
+  val OPTIMIZER_MULTI_LIKE_CONVERSION_THRESHOLD =
+    buildConf("spark.sql.optimizer.multiLikeConversionThreshold")
       .internal()
-      .doc("Configure the maximum size of the pattern sequence in like all. Spark will convert " +
-        "the logical combination of like to avoid StackOverflowError. 200 is an empirical value " +
-        "that will not cause StackOverflowError.")
+      .doc("Configure the maximum size of the pattern sequence in like [all|any|some]. Spark " +
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:222]: Shall we capitalize: `like [all|any|some]` ->`LIKE [ALL|ANY|SOME]`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -216,16 +216,16 @@ object SQLConf {
...
+        "will convert the logical combination of like to avoid StackOverflowError. 200 is an " +
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:223]: `like` -> `LIKE`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -216,16 +216,16 @@ object SQLConf {
...
+        "empirical value that will not cause StackOverflowError.")
       .version("3.1.0")
       .intConf
       .checkValue(threshold => threshold >= 0, "The maximum size of pattern sequence " +
-        "in like all must be non-negative")
+        "in like [all|any|some] must be non-negative")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:228]: `like [all|any|some]` -> `LIKE [ALL|ANY|SOME]`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 3 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1410]: Yeah

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 4 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:222]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 5 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:223]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 6 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:228]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1399]: Let's remove `getLikeQuantifierExprs`

**Comment by beliefer** [comment]
retest this please

**Comment by wangyum** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 7]
```diff
@@ -1404,12 +1396,26 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
...
+              ctx.expression.asScala.map(expression)
+                .map(p => invertIfNotDefined(new Like(e, p))).toSeq.reduceLeft(Or)
+            }
           case Some(SqlBaseParser.ALL) =>
             validate(!ctx.expression.isEmpty, "Expected something between '(' and ')'.", ctx)
-            val expressions = ctx.expression.asScala.map(expression)
-            if (expressions.size > SQLConf.get.optimizerLikeAllConversionThreshold &&
-              expressions.forall(_.foldable) && expressions.forall(_.dataType == StringType)) {
+            val expressions = expressionList(ctx.expression)
+            if (expressions.size > 200 && expressions.forall(_.foldable) &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1417]: do we need this `expressions.size > 200` condition? Seems fine if we always go to the LIKE ALL/ANY version?

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:1417]: Yeah

**Comment by beliefer** [comment]
retest this please

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan Thanks for your work!
@wangyum @dongjoon-hyun Thanks for your review.

**Pull request closed without merging**